README.md
GoSSHa
=================================

[![Build Status](https://travis-ci.org/vodolaz095/gossha.svg)](https://travis-ci.org/vodolaz095/gossha)
[![GoDoc](https://godoc.org/github.com/vodolaz095/gossha?status.svg)](https://godoc.org/github.com/vodolaz095/gossha)

[Руководство пользователя на Русском языке](https://github.com/vodolaz095/gossha/blob/master/README_RU.md)

Cross-platform ssh-server based chat program, with data persisted into
relational databases of MySQL, PostgreSQL or Sqlite3.
Public channel (with persisted messages) and private message (not stored) are supported.
Application has serious custom scripting and hacking potential.

Use case - devops chat with possibility to run scripts from chat, without SSH access to server.


Main addvantages
=================================

1) [Secure SHell protocol](https://en.wikipedia.org/wiki/Secure_Shell) is used to make all communications safe and secure.

2) Users' profiles and messages are stored in relational databases in easy to manipulate format, so we can use 3rd party applications to work with them.

3) Users can be authorized by passwords or private keys.

4) We can start application listening on few addresses and ports on the same time. For example, listeing on `192.168.1.2:2222` on local area network, and on `193.41.32.25:27015` for uplink connections.

5) Users can execute scripts defined by admin on behalf of local system user running the GoSSHa server.

6) Application can run scripts after each public or private message, with senders name, ip, message exported as environment variables. See `homedir/` folder for examples

7) Application is created in `Go` language, and can be build on many environments and architectures - `Linux`, `Microsoft Windows`, `MacOs`.

Usage
=================================

Firstly, you can create admin account by calling `$ gossha root [username]`

```shell

	[vodolaz095@rhel ~]$ gossha root admin
	  ____      ____ ____  _   _
	 / ___| ___/ ___/ ___|| | | | __ _
	| |  _ / _ \___ \___ \| |_| |/ _` |
	| |_| | (_) |__) |__) |  _  | (_| |
	 \____|\___/____/____/|_| |_|\__,_|


	Persistent SSH based chat for the ones, who cares.
	Build: 1.24.1.b06789e.Linux.x86_64
	Version: Build #b06789e on rhel.Linux.x86_64 on Sun Jun 28 01:10:39 MSK 2015

	Console commands avaible:
	 $ gossha ban [username] - delete user and all his/her messages
	 $ gossha passwd [username] - create/update ordinary user by name and password
	 $ gossha root [username] - create/update root user by name and password

	Empty argument - start in server mode

	Enter password:
	User admin is created and/or new password is set!

```

Than you can login using any of [SSH clients](https://en.wikipedia.org/wiki/Comparison_of_SSH_clients)

For example, like this

```shell

		$ ssh admin@localhost -p 27015

```

Than you can import you private ssh key to be used instead of password by using
the `\k` command.

```

	[vodolaz095@rhel ~]$ ssh admin@localhost -p 27015
	Host key fingerprint is 3d:63:45:c4:82:03:ca:99:80:49:03:8e:f2:d8:3a:bb
	+--[ RSA 2048]----+
	|+=.   .. . oo    |
	|= .o +  o ...    |
	|o.  =    . ..    |
	|.+       . .     |
	|. o     S =      |
	| .       . o     |
	|o                |
	| o               |
	|E.               |
	+-----------------+

	admin@localhost's password:
	GoSSHa - very secure chat.
	Build #1.24.1.b06789e.Linux.x86_64
	Version: Build #b06789e on rhel.Linux.x86_64 on Sun Jun 28 01:10:39 MSK 2015
	Commands avaible:
	 \b - (B)an user (you need to have `root` permissions!)
	 \e - Close current session
	 \exit - Close current session
	 \f - (F)orgot localy available SSH key used for authorising your logins via this client
	 \h - (H)elp, show this screen
	 \i - Print (I)nformation about yourself
	 \k - Use locally available SSH (K)eys to authorise your logins on this server
	 \passwd - Changes current user password
	 \q - Close current session
	 \quit - Close current session
	 \r - (R)egister new user (you need to have `root` permissions!)
	 \rr - (R)egister new (r)oot user (you need to have `root` permissions!)
	 \w - List users, (W)ho are active on this server
	 \x - E(X)ecutes custom user script from home directory
	 all other input is treated as message, that you send to server


	[admin@localhost.localdomain(127.0.0.1) x]{14:14:56}:hello!!!
	[admin@localhost.localdomain(127.0.0.1) *]{02:24:04}:\k
	Importing public key...
	Key imported succesefully!
	[admin@localhost.localdomain(127.0.0.1) *]{02:24:04}:

```
Ordinary messages are colored in `white`, system messages - `green`, private
messages - `blue`.
To send private message, type `@`, than username (`TAB` autocompletion works) to
whom you want to send private message of the record. Private messages are not
stored in the database, and they disapear, when user logouts and logins.


Configuration parameters
=================================
Application can be configured in few wayes (ordered by priority).

1) By starting application with flags defined.

2) By environment variables

3) By JSON object values in config file loaded from `/etc/gossha/gossha.json`

4) By JSON object values in config file loaded from `$HOME/.gossha/gossha.json`

This is example config file provided with application:

```toml

# Automatically generated config file for GoSSHa - SSH powered chat
# Place it either in
#   /etc/gossha/gossha.toml
# or
#   ~/.gossha/gossha.toml
#

# Enable debug
Debug=true

# On what port to listen for all interfaces (like for 0.0.0.0 address)
Port = 27015

# What addresses to bind to
BindTo = ["127.0.0.1:27014"]

#Setting database connection - various possible combinations are shown

#SQLite3 with database in local file
#Driver = "sqlite3"
#ConnectionString = "/home/vodolaz095/.gossha/gossha.db"

#SQLite3 with database in memory
#Driver = "sqlite3"
#ConnectionString = ":memory:"

#MySQL database
#Driver = "mysql"
#ConnectionString = "username:password@hostname/database?charset=utf8&parseTime=True&loc=Local"

#PostgreSQL database. 1st variant
#Driver = "postgres"
#ConnectionString ="user=gorm dbname=gorm sslmode=disable"

#PostgreSQL database. 2nd variant
#Driver="postgres"
#ConnectionString="postgres://pqgotest:password@localhost/pqgotest?sslmode=verify-full")"

#This database connection setting are being used
Driver = "sqlite3"
ConnectionString = "/home/vodolaz095/.gossha/gossha.db"


#Path to SSH Public key
SshPublicKeyPath = "/home/vodolaz095/.ssh/id_rsa.pub"
#Path to SSH Private key
SshPrivateKeyPath = "/home/vodolaz095/.ssh/id_rsa"

#Directory to search for custom scripts
Homedir = "/home/vodolaz095/.gossha"

#Script to be executed on each message
ExecuteOnMessage=""

#Script to be execute on each private message
ExecuteOnPrivateMessage=""


```

***Port*** (integer) for application to listein on `0.0.0.0` address (all interfaces). The
default value is `27015`, it can be via `GOSSHA_PORT=27015`
environment value.


***Debug*** (boolean) toggle mode with usage of more verbose output to stdout and start [pprof](https://golang.org/pkg/net/http/pprof/)
server on [http://localhost:6060](http://localhost:6060) port for debugging/benchmarking purposes.
Can be enabled by `GOSSHA_DEBUG=true` environment value.

***Driver*** and ***connectionString*** sets the connection to database.
We can use [sqlite3](https://github.com/mattn/go-sqlite3),
[MySQL](https://github.com/go-sql-driver/mysql) (`MariaDB` in compatibility mode),
[PostgreSQL](https://github.com/lib/pq) databases via appropriate drivers.

Possible pairs of values are

```
   	--driver=sqlite3 --connectionString=/var/lib/gossha/gossha.db

   	--driver=sqlite3 --connectionString=:memory:

   	--driver=mysql --connectionString='user:password@/dbname?charset=utf8&parseTime=True&loc=Local'

   	--driver=postgres --connectionString='user=gorm dbname=gorm sslmode=disable'

   	--driver=postgres --connectionString='postgres://pqgotest:password@localhost/pqgotest?sslmode=verify-full'

```
by default, the `sqlite3` driver is used with database stored at `$HOME/.gossha/gossha.db`.
We can load `driver` from `GOSSHA_DRIVER=sqlite3` environment value and load `connectionString` from `GOSSHA_CONNECTIONSTRING=:memory:` environment value

***SshPublicKeyPath*** points to Public Key to be used by SSH server, default value is `$HOME/.ssh/id_rsa.pub`.
Can be set via  `GOSSHA_SSHPUBLICKEYPATH=/home/myusername/.ssh/id_rsa.pub` environment value.

***sshPrivateKeyPath*** points to Private Key to be used by SSH server, default value is `$HOME/.ssh/id_rsa.pub`.
Can be set via `GOSSHA_SSHPRIVATEKEYPATH=/home/myusername/.ssh/id_rsa` environment value.

***Homedir*** is path to directory containing user's scripts to be executed via `\x` command in chat. It is worth notice,
that this scripts have to be executable files, like the examples, provided in `homedir/scripts` directory of
the distribution or repo. The username, ip and other data is populated from environment values used for scripts.
We can make this executable files in any language - shell, binaries, nodejs files, php scripts.
Can be set by `GOSSHA_HOMEDIR=/home/myusername/.gossha` environment value.

***executeOnMessage*** is path to executable to be started on each message.
We can make this executable files in any language - shell, binaries, nodejs files, php scripts.
See `homedir/afterMessage` for shell example.
Can be set via `GOSSHA_EXECUTEAFTERMESSAGE=/home/myusername/.gossha/afterMessage` environment value.

***executeOnPrivateMessage*** is path to executable to be started on each message.
We can make this executable files in any language - shell, binaries, nodejs files, php scripts.
See `homedir/afterPrivateMessage` for shell example.
Can be set by means of `GOSSHA_EXECUTEAFTERPRIVATEMESSAGE=/home/myusername/.gossha/afterPrivateMessage` environment value.


Building from sources
=================================
I assume you have one of popular `Linux` distros, i don't care about other OSes.

1) [Install Go language](http://golang.org/doc/install) and it's [environment](http://golang.org/doc/code.html#GOPATH) properly. At least `1.4.2` version.

2) Verify you have [GNU Make](https://www.gnu.org/software/make/) at least of
4.0 version.

3) Clone code from repository in appropriate place

```shell

	$ go get github.com/vodolaz095/gossha

```

3) Try to build 

```shell

	$ go get $GOPATH/github.com/vodolaz095/gossha
	$ make

```
The binary file will be created in `build/gossha`

4) Try to install globaly (root password will be asked!) 

```shell

	$ make install

```
This step results in binary generated and placed in `/usr/bin/gossha`.
Also you can uninstall binaries by (root password will be asked!)

```shell

	$ make uninstall

```

5) By default, when you run the application first time, the directory 
with databases, configs and scripts will be created in `$HOME/.gossha/`



Installation via prebuild binaries
=================================

You can get compiled binaries from here
[https://github.com/vodolaz095/gossha/releases](https://github.com/vodolaz095/gossha/releases)

You can verify the signatures via `GPG` or `GPG2`. It have to be something like this:

```shell

		[vodolaz095@vodolaz095 build]$ gpg2 --verify md5sum.txt.sig md5sum.txt
		gpg: Signature made Mon 29 Jun 2015 02:44:13 AM MSK using RSA key ID 994C6375
		gpg: Good signature from "Anatoliy Ostroumov <ostroumov095@gmail.com>" [ultimate]
		gpg:                 aka "[jpeg image of size 2756]" [ultimate]
		gpg:                 aka "[jpeg image of size 3725]" [ultimate]


```

with RSA key ID of `994C6375`!


License
=================================
The MIT License (MIT)

Copyright (c) 2015 Ostroumov Anatolij ostroumov095(at)gmail(dot)com et al.

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# v 2.1.0
Improve password hashing. Introduce golang modules.

- Introduce golang [modules](https://golang.org/ref/mod) for dependency managment.
- Passwords are now hashed using [Argon2](https://github.com/alexedwards/argon2id) algorithm.
- Code can be compiled using Golang 1.14.13 now on centos8 machine.


# v 2.0.0
Newer dependencies. Code strongly refactored.
Breaking changes:

- configuration is stored in `.toml` file with tons of comments
- code simplification
- more unit tests
- possibility to build application with different database drivers being used. For example, build with only MySQL support, or build with MySQL, SQLite3 and PostgreSQL support
- deprecated overriding configuration by flags - it made online documentation very unfriendly, and using file/environment is easier.

Many typos and small issues fixed



# v 1.1.6
Newer dependencies. Build with Go 1.5.3

# v 1.1.5
Ability to set password for users from shell by calling

```shell

	$ gossha passwd user password
	$ gossha root user password

```

Many typos fixed. Dockerfile provided.

# v 1.1.4
Use other library to conceal password input for creating users
Newer dependencies. Small fixes.

# v 1.1.3
More recent dependencies. Build with Go 1.5.1.

# v 1.1.2
More recent dependencies, build with go1.4.2 linux/amd64

# v 1.1.1
More recent dependencies

# v 1.1.0
Tons of smallfixes, console interface is refactored. Console commands of `gossha list`,`gossha log`,`gossha dumpcfg` are added.
More verbose error reporter with link to bug-tracker.

# v 1.0.4
Newer crypto and gorm libs. More standart and automated `Makefile` behaviour. Readme updated.

# v 1.0.3
Newer crypto and gorm libs

# v 1.0.2
Newer crypto libs

# v 1.0.1
Code style, fixes in continious integration scripts.
Proper build environment for Raspberry Pi v1.
Removed some debug comments

# v 1.0.0
First release candidate


cli/cli.go
package cli

import (
	"encoding/base64"
	"fmt"
	"net/http"
	_ "net/http/pprof" //so we can have debugging on localhost:3000 - See http://godoc.org/net/http/pprof
	"os"
	"strconv"
	"strings"

	"github.com/howeyc/gopass"
	"github.com/spf13/cobra"
	"github.com/vodolaz095/gossha/config"
	"github.com/vodolaz095/gossha/models"
	"github.com/vodolaz095/gossha/ssh"
	"github.com/vodolaz095/gossha/version"
)

// Greet writes some motivating text alongside with application version
func Greet() string {
	g := "ICBfX19fICAgICAgX19fXyBfX19fICBfICAgXyAgICAgICAKIC8gX19ffCBfX18vIF9fXy8gX19ffHwgfCB8IHwgX18gXyAKfCB8ICBfIC8gXyBcX19fIFxfX18gXHwgfF98IHwvIF9gIHwKfCB8X3wgfCAoXykgfF9fKSB8X18pIHwgIF8gIHwgKF98IHwKIFxfX19ffFxfX18vX19fXy9fX19fL3xffCB8X3xcX18sX3wKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAK"
	gg, _ := base64.StdEncoding.DecodeString(g)
	var ggg []string
	ggg = append(ggg, "")
	ggg = append(ggg, string(gg))
	ggg = append(ggg, "Cross-platform ssh-server based chat program.")
	ggg = append(ggg, fmt.Sprintf("Build: %v", version.Version))
	ggg = append(ggg, "Homepage: https://github.com/vodolaz095/gossha")
	ggg = append(ggg, "Error reporting: https://github.com/vodolaz095/gossha/issues")
	ggg = append(ggg, "API documentation: https://godoc.com/github.com/vodolaz095/gossha")
	ggg = append(ggg, "           ")
	return strings.Join(ggg, "\r\n")
}

//ProcessConsoleCommand is a dispatcher for processing console commands and main entry point for application
func ProcessConsoleCommand() {
	var rootCmd = &cobra.Command{
		Use:   "gossha",
		Short: "GoSSHa is a cross-platform ssh-server based chat program",
		Long:  "GoSSHa is a cross-platform ssh-server based chat program, with data persisted into relational databases of MySQL, PostgreSQL or Sqlite3. Public channel (with persisted messages) and private message (not stored) are supported. Application has serious custom scripting and hacking potential.",
		Run: func(cmd *cobra.Command, args []string) {
			fmt.Println(Greet())
			fmt.Println()
			fmt.Println("Try `gossha help` for help...")
			fmt.Println()
			if config.RuntimeConfig.Debug {
				fmt.Println("Debug server is listening on http://localhost:3000/debug/pprof!")
				go func() {
					fmt.Println(http.ListenAndServe("localhost:3000", nil))
				}()
			}
			if len(config.RuntimeConfig.BindTo) > 0 {
				for _, v := range config.RuntimeConfig.BindTo[:(len(config.RuntimeConfig.BindTo) - 1)] {
					go ssh.StartSSHD(v)
				}
				ssh.StartSSHD(config.RuntimeConfig.BindTo[len(config.RuntimeConfig.BindTo)-1])
			} else {
				ssh.StartSSHD(fmt.Sprintf("0.0.0.0:%v", config.RuntimeConfig.Port))
			}
		},
	}
	var versionCmd = &cobra.Command{
		Use:   "version",
		Short: "Outputs program version and exits",
		Long:  "Outputs program version and exits",
		Run: func(cmd *cobra.Command, args []string) {
			fmt.Println(Greet())
			os.Exit(0)
		},
	}
	var passwdCmd = &cobra.Command{
		Use:   "passwd [username] [password]",
		Short: "Creates user or set new password to existent one",
		Long:  "Creates user or set new password to existent one",
		Run: func(cmd *cobra.Command, args []string) {
			switch len(args) {
			case 2:
				name := args[0]
				password := args[1]
				err := models.CreateUser(name, password, false)
				if err != nil {
					panic(err)
				}
				fmt.Printf("User %v is created and/or new password is set!\n", name)
				os.Exit(0)
				break
			case 1:
				name := args[0]
				fmt.Print("Enter password:")
				password, err := gopass.GetPasswd()
				if err != nil {
					panic(err)
				}
				err = models.CreateUser(name, string(password), false)
				if err != nil {
					panic(err)
				}
				fmt.Printf("User %v is created and/or new password is set!\n", name)
				os.Exit(0)
				break
			default:
				fmt.Printf("Enter user's name!\n")
				os.Exit(1)
			}
		},
	}
	var makeRootUserCmd = &cobra.Command{
		Use:   "root [username] [password]",
		Short: "Creates root user or set new password to existent one",
		Long:  "Creates root user or set new password to existent one",
		Run: func(cmd *cobra.Command, args []string) {
			switch len(args) {
			case 2:
				name := args[0]
				password := args[1]
				err := models.CreateUser(name, password, true)
				if err != nil {
					panic(err)
				}
				fmt.Printf("Admin %v is created and/or new password is set!\n", name)
				os.Exit(0)
				break
			case 1:
				name := args[0]
				fmt.Print("Enter password:")
				password, err := gopass.GetPasswd()
				if err != nil {
					panic(err)
				}
				err = models.CreateUser(name, string(password), true)
				if err != nil {
					panic(err)
				}
				fmt.Printf("Admin %v is created and/or new password is set!\n", name)
				os.Exit(0)
				break
			default:
				fmt.Printf("Enter user's name!\n")
				os.Exit(1)
			}
		},
	}
	var banCmd = &cobra.Command{
		Use:   "ban [username]",
		Short: "Delete user and all his/her messages",
		Long:  "Delete user and all his/her messages",
		Run: func(cmd *cobra.Command, args []string) {
			if len(args) == 1 {
				name := args[0]
				err := models.BanUser(name)
				if err != nil {
					panic(err)
				}
				fmt.Printf("User %v is banned!\n", name)
				os.Exit(0)
			} else {
				fmt.Printf("Enter user's name!\n")
				os.Exit(1)
			}
		},
	}
	var dumpConfig = &cobra.Command{
		Use:   "dumpcfg",
		Short: "Outputs the configuration in YAML format",
		Long:  "Outputs the configuration in YAML format. Save this config in `$HOME/.gossha/gossha.toml` or `/etc/gossha/gossha.toml`",
		Run: func(cmd *cobra.Command, args []string) {
			json, err := config.RuntimeConfig.Dump()
			if err != nil {
				panic(err)
			}
			fmt.Println(json)
		},
	}

	var listUsers = &cobra.Command{
		Use:   "list",
		Short: "List users",
		Long:  "List users",
		Run: func(cmd *cobra.Command, args []string) {
			var users []models.User
			k := 0
			err := models.DB.Table("user").Order("user.name ASC").Find(&users).Error
			if err != nil {
				panic(err)
			}
			fmt.Println("Users in database:")
			for _, u := range users {
				k++
				if u.Root {
					fmt.Printf("%v) %v (root!) - online on %v \n", k, u.Name, u.LastSeenOnline.Format("2006-1-2 15:04:05"))
				} else {
					fmt.Printf("%v) %v - online on %v \n", k, u.Name, u.LastSeenOnline.Format("2006-1-2 15:04:05"))
				}
			}
		},
	}

	var listMessages = &cobra.Command{
		Use:   "log [limit]",
		Short: "Show last messages, default limit is 10",
		Long:  "Show last messages, default limit is 10",
		Run: func(cmd *cobra.Command, args []string) {
			var ret []models.Notification
			var messages []models.Message
			var limit int
			if len(args) == 1 {
				l, _ := strconv.ParseInt(args[0], 10, 8)
				if l > 0 {
					limit = int(l)
				} else {
					limit = 10
				}
			} else {
				limit = 10
			}
			err := models.DB.Table("message").Preload("User").Limit(limit).Order("message.id desc").Find(&messages).Error
			if err != nil {
				panic(err)
			}
			for _, m := range messages {
				ret = append(ret, models.Notification{User: m.User, Message: m, IsSystem: false, IsChat: true})
			}
			for _, n := range ret {
				var u = n.User
				var m = n.Message
				var online string
				if u.IsOnline() {
					online = "*"
				} else {
					online = "x"
				}
				fmt.Printf("[%v@%v(%v) %v]{%v}:%v\r\n", u.Name, m.Hostname, m.IP, online, m.CreatedAt.Format("2006-1-2 15:04:05"), m.Message)
			}
		},
	}

	//Note! - this flags are actually used in `config.go#InitConfig`.
	//They are copied here to make application more user friendly!
	/*
		rootCmd.PersistentFlags().Uint("port", 27015, "set the port to listen for connections")
		rootCmd.PersistentFlags().Bool("debug", false, "start pprof debugging on http://localhost:3000/debug/pprof/. See `http://godoc.org/net/http/pprof`")
		rootCmd.PersistentFlags().String("driver", "sqlite3", "set the database driver to use, possible values are `sqlite3`,`mysql`,`postgres`")
		rootCmd.PersistentFlags().String("connectionString", GetDatabasePath(), MakeDSNHelp())
		rootCmd.PersistentFlags().String("sshPublicKeyPath", GetPublicKeyPath(), "location of public ssh key to be used with server, usually the $HOME/.ssh/id_rsa.pub")
		rootCmd.PersistentFlags().String("sshPrivateKeyPath", GetPrivateKeyPath(), "location of private ssh key to be used with server, usually the $HOME/.ssh/id_rsa")
		rootCmd.PersistentFlags().String("homedir", GetHomeDir(), "The home directory of module, usually $HOME/.gossha")
		rootCmd.PersistentFlags().String("executeOnMessage", "", "Script to execute on each message")
		rootCmd.PersistentFlags().String("executeOnPrivateMessage", "", "Script to execute on each private message")
	*/
	rootCmd.AddCommand(versionCmd, passwdCmd, makeRootUserCmd, banCmd, dumpConfig, listUsers, listMessages)
	rootCmd.Execute()
}
